packages/dds/tree/src/shared-tree/sharedTree.ts

```diff
@@ -95,14 +94,73 @@ export interface ISharedTree extends ICheckout<IDefaultEditBuilder>, ISharedObje
 	 * Something should ensure the document contents are always in schema.
 	 */
 	readonly storedSchema: StoredSchemaRepository;
+	/**
+	 * Current contents.
+	 * Updated by edits (local and remote).
+	 * Use `runTransaction` to create a local edit.
+	 */
+	readonly forest: IForestSubscription;
+
+	/**
+	 * Run `transaction` to edit this forest.
+	 * While `transaction` is running, its intermediate states will be visible on the IForestSubscription.
+	 *
+	 * TODO: support nesting (perhaps via "commands"),
+	 * and do this in a way where there is control over which ones intermediate versions are displayed.
+	 */
+	runTransaction(
+		transaction: (
+			forest: IForestSubscription,
+			editor: IDefaultEditBuilder,
+		) => TransactionResult,
+	): TransactionResult;
+
+	/**
+	 * Spawn a new checkout which is based off of the current state of this checkout.
+	 * Any mutations of the new checkout will not apply to this checkout until the new checkout is merged back in.
+	 */
+	fork(): ISharedTreeCheckoutFork;
 }
 
+/**
+ * An `ISharedTreeCheckout` which has been forked from a pre-existing checkout.
+ * @alpha
+ */
+export interface ISharedTreeCheckoutFork extends ISharedTreeCheckout {
+	/**
+	 * Rebase the changes that have been applied to this checkout over all the changes in the base checkout that have
+	 * occurred since this checkout last pulled (or was forked).
+	 */
+	pull(): void;
+
+	/**
+	 * Apply all the changes on this checkout to the base checkout from which it was forked. If the base checkout has new
+	 * changes since this checkout last pulled (or was forked), then this checkout's changes will be rebased over those first.
+	 * After the merge completes, this checkout may no longer be forked or mutated.
+	 */
+	merge(): void;
+
+	/**
+	 * Whether or not this checkout has been merged into its base checkout via `merge()`.
+	 * If it has, then it may no longer be forked or mutated.
+	 */
+	isMerged(): boolean;
```

Comment:
Why do we need this policy that branches can't be edited after merging?

Reply:
It's to prevent the user from forgetting to dispose(). Right now we don't actually dispose anything, but eventually there might be indexes or edit manager state or something that needs to be cleaned up when the branch is all done being used. If a user can have a "long lived" branch that repeatedly merges into its base, it seems it would be easy to forget to dispose it. That said, if we find in the future that "long lived" branches would be very useful, I think the code is already structured such that it'd be easy to change this to allow that.